lib/types/record.nix

```diff
@@ -0,0 +1,86 @@
+{ lib }:
+
+let
+  inherit (lib)
+    mapAttrs zipAttrsWith removeAttrs attrNames showOption
+    mergeDefinitions mkOptionType isAttrs
+    ;
+
+  record = args@{ fields, wildcard ? null }:
```

Comment:
What's missing is something like `optionalFields`.
It's best not to mix required fields and optional fields, because
- required fields can be returned without looking at their declarations, making it a bit more lazy, which is good for performance and robustness
- if we don't have optional or wildcard fields, we can return all the fields without even looking at the definitions
- it avoids having to partition the required and optional fields at runtime, which we'd then have to do; it's more efficient this way